### ISSUE TITLE
[5.2]: Topology test is flaky and failing gating

### ISSUE THREAD

**Issue by yaronkaikov** [scylladb/scylladb#14127]
3 out of the last 4 runs failed on `test_topology` each time on a different mode. This is preventing for promotion next-5.2
https://jenkins.scylladb.com/job/scylla-5.2/job/next/155/testReport/junit/(root)/test_topology/Unit_Tests___test_nodes_with_different_smp/

```
manager = <test.pylib.manager_client.ManagerClient object at 0x7fbf452a7710>
random_tables = <test.pylib.random_tables.RandomTables object at 0x7fbf44114110>

    @pytest.mark.asyncio
    async def test_nodes_with_different_smp(manager: ManagerClient, random_tables: RandomTables) -> None:
        # In this test it's more convenient to start with a fresh cluster.
        # We don't need the default nodes,
        # but there is currently no way to express this in the test infrastructure
    
        # When the node starts it tries to communicate with others
        # by sending group0_peer_exchange message to them.
        # This message can be handled on arbitrary shard of the target node.
        # The method manager.server_add waits for node to start, which can only happen
        # if this message has been handled correctly.
        #
        # Note: messaging_service is initialized with server_socket::load_balancing_algorithm::port
        # policy, this means that the shard for message will be chosen as client_port % smp::count.
        # The client port in turn is chosen as rand() * smp::count + current_shard
        # (posix_socket_impl::find_port_and_connect).
        # If this succeeds to occupy a free port in 5 tries and smp::count is the same
        # on both nodes, then it's guaranteed that the message will be
        # processed on the same shard as the calling code.
        # In the general case, we cannot assume that this same shard guarantee holds.
        logger.info(f'Adding --smp=3 server')
        await manager.server_add(cmdline=['--smp', '3'])
    
        # Remove the original 3 servers, the problem is easier to reproduce with --smp values
        # that we pick, not the (currently) default --smp=2 coming from the suite.
        logger.info(f'Decommissioning old servers')
        servers = await manager.running_servers()
        for s in servers[:-1]:
>           await manager.decommission_node(s.server_id)

test/topology/test_topology.py:186: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/pylib/manager_client.py:183: in decommission_node
    await self.client.get_text(f"/cluster/decommission-node/{server_id}")
test/pylib/rest_client.py:82: in get_text
    ret = await self._fetch("GET", resource_uri, response_type = "text", host = host,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <test.pylib.rest_client.UnixRESTClient object at 0x7fbf452a48d0>
method = 'GET', resource = '/cluster/decommission-node/347'
response_type = 'text', host = None, port = None, params = None, json = None
timeout = None

    async def _fetch(self, method: str, resource: str, response_type: Optional[str] = None,
                     host: Optional[str] = None, port: Optional[int] = None,
                     params: Optional[Mapping[str, str]] = None,
                     json: Optional[Mapping] = None, timeout: Optional[float] = None) -> Any:
        # Can raise exception. See https://docs.aiohttp.org/en/latest/web_exceptions.html
        assert method in ["GET", "POST", "PUT", "DELETE"], f"Invalid HTTP request method {method}"
        assert response_type is None or response_type in ["text", "json"], \
                f"Invalid response type requested {response_type} (expected 'text' or 'json')"
        # Build the URI
        port = port if port else self.default_port if hasattr(self, "default_port") else None
        port_str = f":{port}" if port else ""
        assert host is not None or hasattr(self, "default_host"), "_fetch: missing host for " \
                "{method} {resource}"
        host_str = host if host is not None else self.default_host
        uri = self.uri_scheme + "://" + host_str + port_str + resource
        logging.debug(f"RESTClient fetching {method} {uri}")
    
        client_timeout = ClientTimeout(total = timeout if timeout is not None else 300)
        async with request(method, uri,
                           connector = self.connector if hasattr(self, "connector") else None,
                           params = params, json = json, timeout = client_timeout) as resp:
            if resp.status != 200:
                text = await resp.text()
>               raise HTTPError(uri, resp.status, params, json, text)
E               test.pylib.rest_client.HTTPError: HTTP error 500, uri: http+unix://api/cluster/decommission-node/347, params: None, json: None, body:
E               HTTP error 500, uri: http://127.54.119.65:10000/storage_service/decommission, params: None, json: None, body:
E               {"message": "std::runtime_error (data is currently moving to this node; unable to leave the ring)", "code": 500}

test/pylib/rest_client.py:69: HTTPError
```

**Comment by DoronArazii** [comment]
@kostja please have a look

**Comment by eliransin** [comment]
@DoronArazii this issue lacks priority and it seams you have already decided on an assignee.
If it doesn't have some priority I suspect this issue will never move forward. Also adding CI stability label.

**Comment by DoronArazii** [comment]
@kostja any updates? did you had some initial assessment?

**Comment by kbr-scylla** [comment]
This should be fixed on 5.2 branch by my test backports.
https://github.com/scylladb/scylladb/pull/14215

AFAIU 5.2 branch is being periodically merged to branch-2023.1, there was no merge yet after the above PR got merged into 5.2.
cc @avikivity

**Comment by kostja** [comment]
Per @kbr-scylla messages, closing. Please reopen if the issue is still there.
